icicle/appUtils/poseidon/poseidon.cu

```diff
@@ -0,0 +1,256 @@
+}
+
+__device__ scalar_t vecs_mul_matrix(scalar_t element, scalar_t * matrix, int element_number, int vec_number, int size) {
+    shared_states[threadIdx.x] = element;
+    __syncthreads();
+
+    element = scalar_t::zero();
+    for (int i = 0; i < size; i++) {
+        element = element + (shared_states[vec_number * size + i] * matrix[i * size + element_number]);
+    }
+    __syncthreads();
+    return element;
+}
+
+__device__ scalar_t full_round(scalar_t element,
+                               size_t rc_offset,
+                               int local_state_number,
+                               int element_number,
+                               bool multiply_by_mds,
+                               bool add_round_constant,
+                               const PoseidonConfiguration config) {
+    element = sbox_alpha_five(element);
+    if (add_round_constant) {
+        element = element + config.round_constants[rc_offset + element_number];
+    }
+
+    // Multiply all the states by mds matrix
+    scalar_t * matrix = multiply_by_mds ? config.mds_matrix : config.non_sparse_matrix;
+    return vecs_mul_matrix(element, matrix, element_number, local_state_number, config.t);
+}
+
+// Execute full rounds
+__global__ void full_rounds(scalar_t * states, size_t number_of_states, size_t rc_offset, bool first_half, const PoseidonConfiguration config) {
+    int idx = (blockIdx.x * blockDim.x) + threadIdx.x;
+    int state_number = idx / config.t;
+    if (state_number >= number_of_states) {
+        return;
+    }
+    int local_state_number = threadIdx.x / config.t;
+    int element_number = idx % config.t;
+
+    for (int i = 0; i < config.full_rounds_half - 1; i++) {
+        states[idx] = full_round(states[idx],
+                                 rc_offset,
+                                 local_state_number,
+                                 element_number,
+                                 true,
+                                 true,
+                                 config);
+        rc_offset += config.t;
+    }
+
+    states[idx] = full_round(states[idx],
+                             rc_offset,
+                             local_state_number,
+                             element_number,
+                             !first_half,
+                             first_half,
+                             config);
+}
+
+__device__ scalar_t partial_round(scalar_t * state,
+                                  size_t rc_offset,
+                                  int round_number,
+                                  const PoseidonConfiguration config) {
+    scalar_t element = state[0];
+    element = sbox_alpha_five(element);
+    element = element + config.round_constants[rc_offset];
+
+    scalar_t * sparse_matrix = &config.sparse_matrices[(config.t * 2 - 1) * round_number];
+
+    state[0] = element * sparse_matrix[0];
+    for (int i = 1; i < config.t; i++) {
+        state[0] = state[0] + (state[i] * sparse_matrix[i]);
```

Comment:
The same note about `scalar_t::mul_wide` here

Reply:
Fixed in 24222badb6d0db93bc0f0645ef7103beb4a6fa30